### ISSUE TITLE
Uncaught (in promise) Error: No router instance found. You should only use "next/router" inside the client side of your app.

### ISSUE THREAD

**Issue by rajat1saxena** [zeit/next.js#6713]
# Bug report

## Describe the bug

`next/router` is not working as expected. Instead of redirecting, it is throwing the error mentioned in the title.

## To Reproduce

1. Just setup a new Next.js project by fetching next, react and react-dom.
2. Create an `index.js` page with following code
```
import Router from 'next/router'

export default (props) => {
        Router.push('/hola')

        return  (
            <div>
                Hi
            </div>
        )
}

```
3. It will throw this error

## Expected behavior
It should show the 404 page as 'hola.js' is not there.

## System information
 - OS: Ubuntu 18.10
 - Version of Next.js: 8.0.3

**Comment by MarchWorks** [comment]
No, the error is perfectly fine because the page get rendered in the server if you want this code to work put `Router.push('/hola')` inside `componentDidMount` or access to this page from another page with a Link

**Comment by rajat1saxena** [comment]
If one has to put `Router.push('/hola')` inside `componentDidMount`, how does the following official example work then?
```
import Router from 'next/router'

function ReadMore() {
  return (
    <div>
      Click <span onClick={() => Router.push('/about')}>here</span> to read more
    </div>
  )
}

export default ReadMore
```

**Comment by timneutkens** [comment]
You're **calling** the function inside of render, meaning it's called for every single time `render` is called.

```
<span onClick={() => Router.push('/about')}>here</span>
```

Above means the equivalent of:

```
<span onClick={function() { Router.push('/about') }}>here</span>
```

As you can see it's a function inside of render. Which is only triggered when triggering `onClick` is called, which in turn only happens on the client side.

**Comment by sagar-gavhane** [comment]
Hi,

`useRouter` hook solved my problem. Can you look at Next.js official docs link: https://nextjs.org/docs#userouter

**Comment by mdzakir** [comment]
**Why This Error Occurred**
During SSR you might have tried to access a router method `push`, `replace`, `back`, which is not supported.

**Possible Ways to Fix It**
Move any calls to router methods to `componentDidMount` or add a check such as `typeof window !== 'undefined'` before calling the methods

Source: https://github.com/vercel/next.js/blob/master/errors/no-router-instance.md

**Comment by PedroMarianoAlmeida** [comment]
You can use this workaround:

if (process.browser){
  //Runs only on client side
}

**Comment by prakhar-ap** [comment]
Finally implemented a working solution for this.

First Check whether its a server or not!
`
if(!(typeof window === undefined)) {
      window.history.pushState(null, null, url);
            window.location.reload(true);
}
`

**Comment by huixiangming** [comment]
> 不，错误是完全正常的，因为页面的服务器得到呈现，如果你想这个代码工作放`Router.push('/hola')`里面`componentDidMount`从另一个页面或访问该页面的链接

你这一句话不加标点符号，我读了半天才读懂，我英语不太好

**Comment by definaru** [comment]
If an event occurs in an array of data that you iterate over in a map, then use an `useEffect` with a constant dependency.
This will get rid of the error. Plus arrow function in the `next/router` link. This was used on Next.js

`import React, { useEffect } from 'react'
import { useRouter } from 'next/router'

export default function yourFunc()
{
      const router = useRouter()
      const CustomPanel = [
           {href: '/users', text: 'All Lists user'},
           {href: '/about', text: 'About'}
      ]

    useEffect(() => {}, [CustomPanel]) // All the magic is here

    return (
       <div>
{CustomPanel.map((item, inx) => (
<a key={inx} onClick={() => router.push(`${item.href}`)}>{item.text}</a>
))}
    )
       </div>
}
`

**Comment by GorvGoyl** [comment]
> Hi,
> 
> `useRouter` hook solved my problem. Can you look at Next.js official docs link: https://nextjs.org/docs#userouter

updated link: https://nextjs.org/docs/api-reference/next/router#userouter

**Comment by someshkoli** [comment]
Any way we can invoke this error forcefully ? We are getting this error somewhere in our logs but we don't know the source/cause. Also can this error result in shutting down/restart the app server ? 
We have our application deployed on AWS eb, and while monitoring the application we noticed a lot of BackendConnnectionError on ELB metric, after going through the logs, we found multiple log files in /var/log/containers, majority of which ended with this error. I'm guessing this error is causing the container to restart (resulting in multiple log files)  which is then causing the BackendConnnectionError metics to jump up and ultimately causing latency in the main application. 
Anything to add here ?
